spec14: clarify UC1.3

grondo opened the pull request:
This PR is meant to address the main issue brought up in #105, which I think
everyone agreed was valid and should be fixed in the RFC.

The description of Use Case 1.3 is worded with "at least", but the actual
example jobspec specified the exact count of sockets and cores. Update the
example to use a `min` count instead with no `max` (implying "at least").

Adjust the wording for the task `count` key to allow specification of `min`
without `max`, to make the new UC1.3 adhere to the spec.

Closes #105

dongahn commented:
This looks great me. @grondo: do you want to add the semantics of the spec as to when the plain `count` key is used, instead of its sub-keys in this PR?

SteVwonder commented:
Thanks @grondo!

One thing I noticed, is that (before this PR) UC1.3 and UC1.4 have the exact same jobspec but different text describing them:
```
→ diff use_case_1.3.yaml use_case_1.4.yaml && echo "Files do not differ"
Files do not differ
```
I think UC1.3 is meant to show the node not being exclusively allocated.  Do you mind adding `exclusive: false` under `node` to UC1.3?

Also, if you don't mind, UC2.5 and UC2.6 both have the same wording of `at least` when they should `exactly` (or something similar).  Can you switch from `at least` to `exactly` or change the `count` to use `min`?  Either is fine with me.

grondo commented:
> do you want to add the semantics of the spec as to when the plain count key is used, instead of its sub-keys in this PR?

I'd be happy to -- though I'm not sure exactly what you want. Currently the description states:

> The `count` SHALL have one of two possible values: either a single integer value representing a fixed count, or a dictionary which SHALL contain the following keys

Would you like the case of the "single integer value representing a fixed count" expanded on, or did you have something else in mind?

grondo commented:
>I think UC1.3 is meant to show the node not being exclusively allocated. Do you mind adding exclusive: false under node to UC1.3?

Yes, great catch!

> Also, if you don't mind, UC2.5 and UC2.6 both have the same wording of at least when they should exactly (or something similar). Can you switch from at least to exactly or change the count to use min? Either is fine with me.

Ok, will do.

dongahn commented:
> Would you like the case of the "single integer value representing a fixed count" expanded on, or did you have something else in mind?

Yes, and that plus, what this means to the scheduler in matching and generating the resource set info. If you don't feel comfortable to write this up, I can do this while I'm working on the RFC for R since we will have to link those two RFCs as well.

grondo commented:
> Yes, and that plus, what this means to the scheduler in matching and generating the resource set info

Ah, a sort of *caveat utilitor* clause for the exact `count` specification? That definitely makes sense to be in here, but I'd rather go over that in a separate PR.

grondo commented:
Oops I didn't get UC1.3 right. I replaced `count` with `min` instead of adding `min` as a key under `count`. Fix coming up.

grondo commented:
Ok, fixed, squashed, and forced a push.

>  Can you switch from at least to exactly or change the count to use min? Either is fine with me.

The *at least* in these use cases specified an amount of RAM. I didn't think it would make sense (ever?) to request an *exact* amount of RAM, so I changed the examples to use `min` and kept the wording the same.

garlick commented:
Is this OK to merge?

Note as discussed offline, the example changes need to be propagated to flux-core's jobspec valid input in `t/jobspec/valid`.  I think the jobspec JSON schema also will need updating to allow the dictionary range without `max`.  I'm happy to take care of both but we should get an issue open over there when this is ready to merge.

grondo commented:
I had a working branch that updated the valid inputs and was going to ask you about updating the schema.

It did feel like there should be an option to test against the "upstream" rfc automatically or otherwise keep the valid input synchronized without having to do it manually, but I suppose that would make it more difficult to propose a new revision of jobspec without affecting tests in flux-core.

garlick commented:
Right sounds like there is consensus here so merging.  New PR if anything needs to change.

SteVwonder commented:
@grondo: which version of asciidoc do you have installed?  I'm running into issues with the `[horizontal width=20%]`:

```
→ make
asciidoc -o spec_14.html spec_14.adoc
asciidoc: ERROR: spec_14.adoc: line 135: illegal style name: horizontal width=20%
asciidoc: ERROR: spec_14.adoc: line 141: illegal style name: horizontal width=30%
make: *** [spec_14.html] Error 1

→ asciidoc --version
asciidoc 8.6.8
```

grondo commented:
I don't install asciidoc locally and instead use the GitHub rendering to check rendering. Sorry about that! I guess we should probably add asciidoc to travis-ci to make sure we catch these mistakes before merging. (and now that I look back at this I clearly didn't know what I was doing and GitHub has just fallen back to a normal definition list, which is probably what I wanted anyway in this case)

SteVwonder commented:
Ok.  I'll remove the `width=..` from the horizontal in #154
